connect_four.py

```diff
@@ -101,23 +102,84 @@ def __diagonal_pos_slope(self, row: int, col: int, color: str) -> bool:
         row = self.num_rows - len(self.columns[col])
         return  self.__vertical_win(row, col, color) or self.__horizontal_win(row, col, color) or self.__diagonal_neg_slope(row, col, color) or self.__diagonal_pos_slope(row, col, color)
 
+class InputOutput:
+    def __init__(self, play_mode: bool, moves_list: list) -> None:
+        self.play_mode = play_mode #if TRUE, user input will be asked for. Otherwise, input will be automatic from test functions
+        self.moves_list = moves_list
+        self.move_counter = 0
+
+    def player_names(self) -> str:
+        if self.play_mode:
+            # take player's names
+            print("Player 1, you are yellow. What is your name?")
+            player1 = str(input())
+            print("Player 2, you are red. What is your name?")
+            player2 = str(input())
+        else:
+            player1, player2 = "player1", "player2"
+
+        return player1, player2
+
+    def game_over_IO(self, winner: str, board) -> str:
+        if self.play_mode:
+            print(str(board))
+            if winner:
+                print(f"{winner}, you win!! Press R to restart, or press any key to exit the game.")
+                response = str(input()).capitalize() 
+            else: # __game_over has been called without a winner
+                print(f"The game has ended in a draw. Press R to restart, or press any key to exit the game.")
+                response = str(input()).capitalize()
+        else:
+            response = "Exit"
+        
+        return response
+    
+    def select_col_IO(self, board, player: str) -> int:
+        if self.play_mode:
+            last_col = chr(ord('A') - 1 + board.num_cols)
+
+            print(f"{player}, it is your turn. Enter the column you would like to add a piece to (A-{last_col}):")
+            col = str(input()).capitalize()
+
+            col_code = board.check_col(col)
+            while col_code is not ErrorCode.VALID:
+                if col_code is ErrorCode.COL_INVALID:
+                    print(f"{player}, that column does not exist. Please pick another column (A-{last_col}):") 
+                    col = str(input()).capitalize()
+                    col_code = board.check_col(col)
+                else:
+                    print(f"{player}, that column is full. Please pick another column (A-{last_col}):") 
+                    col = str(input()).capitalize()
+                    col_code = board.check_col(col)
+        
+        else:
+            col_code = None
+            while col_code is not ErrorCode.VALID:
+                col = self.moves_list[self.move_counter].capitalize()
+                self.move_counter += 1 
+                col_code = board.check_col(col)
 
+        return ord(col) - ord('A')
+
+    
 
 # Main gameplay. Set as while loop to play until exited.
 class GameLoop: 
     
-    def __init__(self) -> None:
+    def __init__(self, play_mode: bool = True, moves_list: list = []) -> None:
```

Comment:
I would lean towards using an enum for play mode, because it's not immediately clear to me what play_mode = False means exactly. I would assume it means test mode but it does leave room for ambiguitiy.

That being said, of all the times I told you to use an enum, I feel least strongly about this one, so whatever helps you get to the next stage of incorporating django is most important.